fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,112 @@ where
         Ok((state, res))
     }
 }
+
+impl<DB> FvmMessageInterpreter<DB>
+where
+    DB: Blockstore + 'static + Send + Sync + Clone,
+{
+    fn estimate_gassed_msg(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: &mut Message,
+    ) -> anyhow::Result<Option<GasEstimate>> {
+        // Setting BlockGasLimit as initial limit for gas estimation
+        msg.gas_limit = BLOCK_GAS_LIMIT;
+
+        // estimate the gas limit and assign it to the message
+        // do not reuse the cache
+        let ret = state.call_with_cache(msg.clone(), false)?;
+        if !ret.msg_receipt.exit_code.is_success() {
+            // if the message fail we can't estimate the gas.
+            return Ok(Some(GasEstimate {
+                exit_code: ret.msg_receipt.exit_code,
+                info: ret.failure_info.map(|x| x.to_string()).unwrap_or_default(),
+                gas_limit: 0,
+            }));
+        }
+
+        msg.gas_limit = (ret.msg_receipt.gas_used as f64 * self.gas_overestimation_rate) as u64;
+
+        if msg.gas_premium.is_zero() {
+            // We need to set the gas_premium to some value other than zero for the
+            // gas estimation to work accurately (I really don't know why this is
+            // the case but after a lot of testing, setting this value to zero rejects the transaction)
+            msg.gas_premium = TokenAmount::from_nano(BigInt::from(1));
+        }
+        if msg.gas_fee_cap.is_zero() {
+            // Compute the fee cap from gas premium and applying an additional overestimation.
+            let overestimated_limit = (msg.gas_limit as f64 * self.gas_overestimation_rate) as u64;
+            msg.gas_fee_cap = std::cmp::min(
+                TokenAmount::from_atto(BigInt::from(overestimated_limit)) + &msg.gas_premium,
+                TokenAmount::from_atto(BLOCK_GAS_LIMIT),
+            );
```

Comment:
This looks wrong again:
1. It's not a price but a gas amount (potentially equals the block gas limit!)
2. It's `gas_limit * gas_overestimation_rate ** 2`